net: socket: Add intermediate socket layer for offloading

rlubos opened the pull request:
This PR is a proposal to resolve a currently existing problem on how to choose which socket implementation should be used if multiple offloaded socket implementations are available in the system.

The proposal introduces an intermediate socket implementation, called "socket dispatcher". The sole reason for this implementation is to postpone the actual socket creation until the first actual operation on a socket is performed. This leaves an opening for the application to select a network interface (and thereby associated socket implementation if the interface is offloaded) with `SO_BINDTODEVICE` socket option), effectively allowing the application to choose which socket implementation should be used for that particular socket. In case no network interface is selected, the layer will use the existing mechanism to select the default one based on priority and current filtering rules.

rlubos committed:
net: sockets: Add socket dispatcher

Add an intermediate socket implementation called socket dispatcher. This
layer can be used along with the socket offloading, to postpone the
actual socket creation until a first operation on a socket is executed.

This approach leaves an opening to bind a socket to a particular
offloaded network interface, and thus offloaded socket implementation,
using SO_BINDTODEVICE socket option. Thanks to this, it is now possible
to use multiple offloaded sockets implementations along with native
sockets, and easily select which socket should use with network
interface (even if it's an offloaded interface).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: if: Connect interface with offloaded socket implementation

Instead of keeping a boolean informing whether a network interface is
offloaded at socket layer or not, keep a pointer to a function which
allows to create an offloaded socket. Native interfaces keep this as
NULL, while for offloaded interfaces it allows to connect an offloaded
socket implementation with an interface.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: sockets: Add separate macro for registering offloaded sockets

Add a separate macro for registering offloaded sockets implementation,
along with information in the structure whether the implementation is
offloaded or not. This allows to differentiate between native and
offloaded socket implementations, which is critical for binding socket
API with an interface.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: socket: Add option to create native TLS sock with offloaded TCP

In some cases (for examples when offloaded socket implementation does
not implement TLS functionality) it could be desired to create a native
TLS socket with an underlying offloaded socket.

This cannot be achieved with SO_BINDTODEVICE option only, as TLS socket
type is not really associated with a particular interface - it either
has to be offloaded, or a fully native socket is created (native TLS on
a native interface).

In order to address the problem, introduce TLS_NATIVE socket option.
This option instructs the socket dispatcher layer to create a native TLS
socket. As with the socket dispatcher the underlying socket
implementation is not decided during TLS socket creation, therefore it's
possible to use SO_BINDTODEVICE to choose either native or offloaded
interface for the underlying socket.

Additionally remove NET_SOCKETS_OFFLOAD_TLS Kconfig option, as it's no
longer needed with an runtime option to select whether to offload TLS or
not.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -1077,6 +1079,8 @@ static void modem_net_iface_init(struct net_if *iface)
 			     NET_LINK_ETHERNET);
 	data->net_iface = iface;
+
+	iface->if_dev->socket_offload = offload_socket;
```
Instead of directly manipulating the pointer in iface, an API function in net_if would be more convenient here.

rlubos replied:
Added API function to set/get the pointer.

jukkar commented on subsys/net/lib/sockets/socket_dispatcher.c:
```diff
@@ -0,0 +1,419 @@
+LOG_MODULE_REGISTER(net_sock_dispatcher, CONFIG_NET_SOCKETS_LOG_LEVEL);
+
+/* TODO create actual configs for those */
+#define CONFIG_NET_SOCKET_DISPATCHER_CONTEXT_MAX 4
```
At least earlier there was some checks for preprocessor symbols that started with `CONFIG_` as those are reserved for Kconfig only. I suggest using different name in order to avoid confusion to the reader.

jukkar reviewed (commented):
Cool, just some minor comments. Also some unit tests would be appropriate here.

mniestroj commented on subsys/net/lib/sockets/socket_dispatcher.c:
```diff
@@ -318,6 +328,26 @@ static int sock_dispatch_setsockopt_vmeth(void *obj, int level, int optname,
 			fd = sock_dispatch_native(obj);
 		}
+	} else if ((level == SOL_TLS) && (optname == TLS_NATIVE)) {
+		const int *tls_native = (int *)optval;
+		struct dispatcher_context *ctx = (struct dispatcher_context *)obj;
```
Implicit casting should work without any issues here, as we only try to cast from/to `void *`. IMO it is rather harmful in that case to use explicit casting, as compiler warnings will be suppressed in the case when we change `sock_dispatch_setsockopt_vmeth()` argument types and forgot to update those explicit casts.

rlubos replied:
Dropped the casting.

mniestroj commented on subsys/net/lib/sockets/socket_dispatcher.c:
```diff
@@ -318,6 +328,26 @@ static int sock_dispatch_setsockopt_vmeth(void *obj, int level, int optname,
+			return -1;
+		}
+
+		if (tls_native) {
```
Should it be `if (*tls_native) {` maybe?

What about default, will it be native or non-native? What about passing value of 0, will there be a fallback from non-native to native during socket creation? Same for value of 1, will there be a fallback from native to non-native in case the former is "not supported" (seems that the current code does not assume any fallback, but just trying to confirm)?

rlubos replied:
Fixed. The default depends on the socket priority. Typically, as offloaded sockets are processed first, it means the default will be offloaded socket (provided it accepts TLS socket types.

I wanted to keep the intermadiate "dispatcher" implementation as minimal as poosbile. I'm not sure what fallback is expected here? The option simply allows to enforce creation of a native TLS socket (we can't really use `SO_BINDTODEVICE`, as TLS socket, just as any other other native socket implementation is not bound to any interface). If not used, the dispatcher will rely on socket priorities and filtering function (just as it is right now). If we want to enforce an offloaded TLS creation, we can simply bind the socket to the offloaded interface.

mniestroj replied:
I was thinking about the case when TLS is not supported by offloaded socket implementation, but it (driver implementing offloaded sockets) still is requested to create TLS socket. Should such case be filtered out by calling `->is_supported()` callback, which at this point is not yet implemented correctly for existing offloaded socket drivers (they all return `true` without even checking if that is TLS or not)? Should #44656 fix that?

>I'm not sure what fallback is expected here?

I think I understand the mechanisms better now, so let's forget about that. Before, I was thinking about having a fallback in case some ciphersuite is not supported, but requested to be used, by offloaded socket implementation. That would require to collect lots of socket options (from setsockopt) and passing them to each network socket implementation. Let's stick with the simple approach.

One thing that still bothers me is whether native TLS should be default or offloaded TLS should be default. In theory there is more control over the native TLS stack, easier debug access (to the handshake phase), configuration of ciphersuites, etc. I would be weird if all applications relying on native TLS would now need to call `setsockopt()` with `TLS_NATIVE`. It is one step back from portability IMO. Could we introduce a Kconfig option to "force" native TLS? What do you think about that idea?

jukkar replied:
> One thing that still bothers me is whether native TLS should be default or offloaded TLS should be default.

I also think the native TLS should be the default. Not sure if we need a kconfig option to force it if it is the default one. IMHO the user should do some coding (either adding some options in the code or setting kconfig options etc) in order to use the offload version.

rlubos replied:
> Should https://github.com/zephyrproject-rtos/zephyr/pull/44656 fix that?

This is definitely a valuable addition to existing offloaded implmentations.

rlubos replied:
The default choice is actually not really hardcoded in the networking stack with this proposal. It all depends on the socket priority that the offloaded implementation is registered with:
* if the offloaded socket have higher priority than the native one, and the offloaded implementation supports TLS, then this will be the default choice for the TLS sockets,
* if the offloaded socket have lower priority than the native one, the default will be native socket, no further action is needed by the application.

So the default choice is actually controlled by the offloaded engine provider. For example, for nRF91 and NCS examples we really prefer to default to the offloaded TLS socket, as use of the native TLS stack with nRF91 is actually an exception.

> I would be weird if all applications relying on native TLS would now need to call setsockopt() with TLS_NATIVE.

That's not really the case. `TLS_NATIVE` option is only needed if the offloaded engine supports TLS **and** has higher priority than the native one. I don't think any existing application would be impacted, as currently we don't even support "officially" native TLS socket with offloading (we only have workarounds in NCS, that were rejected upstream some time ago).

If the dispatcher is disabled, or the offloaded engine provider has decided for native sock to be the default by lowering priority, you don't need to use this option.

mniestroj replied:
>The default choice is actually not really hardcoded in the networking stack with this proposal. It all depends on the socket priority that the offloaded implementation is registered with

That is actually what I figured out yesterday, so thanks for confirming that!

>So the default choice is actually controlled by the offloaded engine provider. For example, for nRF91 and NCS examples we really prefer to default to the offloaded TLS socket, as use of the native TLS stack with nRF91 is actually an exception.

As you say, this is what **NCS examples** prefer. This is reasonable, as those should show the use case of offloaded TLS and its advantages.

However, for me as application / networking stack developer, I want to write as generic code as possible. This means that native TLS will be the default choice, because:
- it works the same on all platforms (compared to offloaded TLS that is platform/driver/networking chip specific),
- I will configure mbedTLS with supported ciphersuite and all other bits, as most platforms will still require it; I can do it once for all platforms, instead of bothering with each offloaded TLS one by one (configuration of offloaded TLS is usually limited, but it is still required to verify if ciphersuite is compatible and test if it actually works),
- native TLS allows to inspect TLS handshake and debug problems if some occur; debugging TLS is hard, but at least it is possible with open solution, such as mbedTLS (under Zephyr native TLS implementation)

IMHO open solution such as native TLS (with full software stack being open source) should be promoted by open projects such as Zephyr.

I have created related PR #44685, which allows to configure offloaded sockets priorities. So far I did not change priority value, so offloaded sockets (and offloaded TLS) is prioritized over native TLS sockets. In order to prioritize native TLS we can just swap priority values, which I think we should do (for the reasons I described above).

nRF91 examples can either increase priority of nRF91 modem offloaded sockets or introduce driver specific option that overwrites "Zephyr default" offloaded socket priority.

>If the dispatcher is disabled, or the offloaded engine provider has decided for native sock to be the default by lowering priority, you don't need to use this option.

Makes sense, thanks!

rlubos replied:
Some generalization in terms of offloaded sockets priority makes sense I guess, thanks for the PR.

mniestroj commented on include/net/net_if.h:
```diff
@@ -454,7 +454,7 @@ struct net_if_dev {
 
 #if defined(CONFIG_NET_SOCKETS_OFFLOAD)
 	/** Indicate whether interface is offloaded at socket level. */
-	bool offloaded;
+	int (*socket_offload)(int family, int type, int proto);
```
Comment needs to be updated. Also the name `socket_offload` does not suggest that this is for socket creation, so maybe a more descriptive name could be used.

rlubos replied:
Updated the comment. I'm not sure about the naming, IMO `socket_offload` is quite straightforward (just like regular `socket()` but with a suffix indicating it refers to an offloaded interface). Do you have any suggestions regarding an alternative name?

mniestroj replied:
`socket()` is known enough to get idea what it does (socket creation). It was not that clear to me when seeing `->socket_offload(family, type, proto);` callback invocation. Was thinking about `socket_offload_create`, but not sure that improves something.

mniestroj commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -166,6 +166,18 @@ config NET_SOCKETS_OFFLOAD
+	help
+	  If enabled, an intermediate socket offloading layer is included
+	  (called socket dispatcher), allowing to select an offloaded network
+	  interace and thus socket implementation with SO_BINDTODEVCIE socket
```
s/interace/interface/

mniestroj reviewed (commented):
I like the idea!

carlescufi commented:
CC @johnlange2

jukkar commented on include/net/socket.h:
```diff
@@ -145,6 +145,18 @@ struct zsock_pollfd {
  */
```
typo in commit msg of `net: socket: Add option to create native TLS sock with offloaded TCP`
`s/implmenetation/implementation/`

jukkar reviewed (commented):
LGTM, minor nit about a typo

rlubos committed:
tests: net: Add socket dispatcher unit tests

Add unit tests for socket dispatcher module. The tests verify that:
* a socket is automatically dispatched to a regular socket on any
  first socket call, if the socket was not bound already before
  (according to the default priority rules)
* a socket is correctly dispatched to an offloaded/native socket
  implementation after a SO_BINDTODEVICE call
* a socket is correctly dispatched to a native TLS socket, with an
  underlying native/offloaded socket, if TLS_NATIVE and SO_BINDOTODEVICE
  socket options are used

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

hubertmis commented on include/net/socket.h:
```diff
@@ -145,6 +145,18 @@ struct zsock_pollfd {
+ *  Note, that this option is only applicable if socket dispatcher is used
+ *  (CONFIG_NET_SOCKETS_OFFLOAD_DISPATCHER is enabled).
+ *  In such case, it should be the first socket option set on a newly created
+ *  socket. After that, the application may use SO_BINDTODEVCIE to choose the
```
nit: typo
```suggestion
 *  socket. After that, the application may use SO_BINDTODEVICE to choose the
```

hubertmis commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -166,14 +166,27 @@ config NET_SOCKETS_OFFLOAD
+	help
+	  If enabled, an intermediate socket offloading layer is included
+	  (called socket dispatcher), allowing to select an offloaded network
+	  interface and thus socket implementation with SO_BINDTODEVCIE socket
```
nit: typo
```suggestion
	  interface and thus socket implementation with SO_BINDTODEVICE socket
```

hubertmis reviewed (commented):
Do I understand correctly that intended usage of this socket dispatcher is something like:
```
struct ifreq ifreq_simplelink = { 
  .ifr_name = "simplelink_dev_label"
};
struct ifreq ifreq_ublox = {
  .ifr_name = "ublox_dev_label"
};

int sock1 = socket(AF_INET6, SOCK_DGRAM, IPPROTO_UDP);
int sock2 = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);

setsockopt(sock1, SOL_SOCKET, SO_BINDTODEVICE, &ifreq_simplelink, sizeof(ifreq_simplelink));
setsockopt(sock2, SOL_SOCKET, SO_BINDTODEVICE, &ifreq_ublox, sizeof(ifreq_ublox));

bind(sock1, ...);
sendto(sock2, ...);
```

rlubos commented:
@hubertmis Thank you for the feedback, yes you're correct about the usage - I plan to add some documentation with an example use.

hubertmis reviewed (dismissed):
Then I like it very much! It's a step in a right direction.

rlubos committed:
doc: net: sockets: Add offloading and dispatcher documentation

Add brief documentation for socket offloading and socket dispatcher.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
I've added unit tests and documentation to the PR, please have a look (especially at the docs).

rlubos commented:
Rebased && resolved conflicts

catch-twenty-two commented:
Thanks for adding the documentation @rlubos! this will be very helpful in our implementation of multiple interfaces

catch-twenty-two commented on doc/connectivity/networking/api/sockets.rst:
```diff
@@ -143,6 +143,137 @@ Secure sockets offer the following options for socket management:
+In case no ``SO_BINDTODEVICE`` socket option is used on a socket, the socket
+will be dispatched according to the default priority and filtering rules on a
+first socket API call.
+
```
Looks great, thanks for adding examples

rerickson1 commented on tests/net/socket/offload_dispatcher/src/main.c:
```diff
@@ -0,0 +1,904 @@
+	bool sendmsg_called;
+	bool getsockname_called;
+	bool getpeername_called;
+} test_socket_ctx[2];
```
Maybe a define for the value `2` here would make the code more clear.

rerickson1 commented on doc/connectivity/networking/api/sockets.rst:
```diff
@@ -143,6 +143,137 @@ Secure sockets offer the following options for socket management:
+   /* A "dispatcher" socket is created */
+   sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+
+	struct ifreq ifreq = {
+		.ifr_name = "SimpleLink"
+	};
+
```
spacing looks off here.

rlubos replied:
Yep, tabs slipped in.

rerickson1 reviewed (commented):
Looks good. A few small things I found.

jukkar commented on include/zephyr/net/socket.h:
```diff
@@ -945,8 +946,18 @@ struct net_socket_register {
+		.is_offloaded = true,					\
 		.is_supported = _is_supported,				\
 		.handler = _handler,					\
 	}
```
As these macros are basically doing the same thing, I was pondering if we could simplify this a bit, so something like this
```
#define _NET_SOCKET_REGISTER(socket_name, prio, _family, _is_supported, _handler, _is_offloaded)  \
...

#define NET_SOCKET_REGISTER(socket_name, prio, _family, _is_supported, _handler) \
	_NET_SOCKET_REGISTER(socket_name, prio, _family, _is_supported, _handler, false)

#define NET_SOCKET_OFFLOAD_REGISTER(socket_name, prio, _family, _is_supported, _handler) \
	_NET_SOCKET_REGISTER(socket_name, prio, _family, _is_supported, _handler, true)
```

rlubos replied:
Good point, I've applied your suggestion.

jukkar reviewed (dismissed):
Just minor nit, LGTM

mniestroj commented on drivers/wifi/simplelink/simplelink.c:
```diff
@@ -212,6 +212,8 @@ static struct net_offload simplelink_offload = {
 	.put	       = NULL,
 };
 
+int simplelink_socket_create(int family, int type, int proto);
+
```
It will be much safer to declare that function in `simplelink_support.h` file, so that compiler/linker will detect if there is difference between declaration and definition.

rlubos replied:
Indeed, I've moved both declarations to respective headers.

mniestroj commented on drivers/wifi/eswifi/eswifi_core.c:
```diff
@@ -378,6 +378,8 @@ static int eswifi_get_mac_addr(struct eswifi_dev *eswifi, uint8_t addr[6])
 	return 0;
 }
 
+int eswifi_socket_create(int family, int type, int proto);
+
```
It will be much safer to declare that function in `eswifi.h` file, so that compiler/linker will detect if there is difference between declaration and definition.

mniestroj reviewed (changes requested):
Looks good, just minor bits to adjust.

mniestroj commented on subsys/net/lib/sockets/CMakeLists.txt:
```diff
@@ -17,13 +17,10 @@ endif()
 
 zephyr_sources_ifdef(CONFIG_NET_SOCKETS_CAN                sockets_can.c)
 zephyr_sources_ifdef(CONFIG_NET_SOCKETS_PACKET             sockets_packet.c)
+zephyr_sources_ifdef(CONFIG_NET_SOCKETS_SOCKOPT_TLS        sockets_tls.c)
```
I think we need an option to disable native TLS unrelated to `CONFIG_NET_SOCKETS_SOCKOPT_TLS` option, so it doesn't occupy space when offloaded TLS is supposed to be used only. But that can be done in separate PR, just expressing my thoughts.

rlubos replied:
Why do you think we need another option for TLS? From my point of view, `CONFIG_NET_SOCKETS_SOCKOPT_TLS` doesn't need to be used by offloading, just as CAN/PACKET arent used by it, so it can be used exclusively to enable native TLS support. The name is quite unfortunate though, IMO we should drop the `SOCKOPT` part...

mniestroj replied:
>The name is quite unfortunate though, IMO we should drop the SOCKOPT part...

Indeed. It is hard to tell what does that option enable or what it SHOULD enable. Is it native TLS (which it does now), TLS credentials storage or rather TLS extensions to `setsockopt()` (which the name suggest, but it isn't the case I think).
